### ISSUE TITLE
Watch Fails on versions > 1.6.1

### ISSUE THREAD

**Issue by kane-mason** [webpack/watchpack#165]
Hi,

I recently discovered my webpack watch had stopped working after reinstalling node modules. After a lot of digging i pinned the issue down to this package. Basically when i use any other version of this package > 1.6.1 the watcher does not keep the process alive anymore. There is no error in the console but the `process.on('exit', ` handler fires.

I am using webpacks node api as follows:

``` 
var compiler = webpack(params);

  if (watch) {
    compiler.watch({
      ignored: /node_modules/
    }, onBuildComplete);
  }
  else {
    compiler.run(onBuildComplete)
  }
```

I thought it was only me but many of my team mates are experiencing the same thing. 

We are watching quite a lot of files (> 800) which might be a contributing factor.

node:  10.21.0
npm: 6.1.0
webpack: 4.43.0
Windows 10 Enterprise

**Comment by evilebottnawi** [comment]
Please create reproducible test repo

**Comment by kane-mason** [comment]
I have tried to create a reproducible test case but am not winning, and i am unable to share our source code.

I could arrange to have a zoom meeting or such and allow you to walk through the code?

I did some more digging and further narrowed it down to chokidar.. If i delete chokidar v3 and force require chokidar v2 then everything works fine

**Comment by evilebottnawi** [comment]
Maybe bug in `chokidar`, without reproducible steps I can't help sorry

**Comment by evilebottnawi** [comment]
No need :+1: please, it is not help, and will be ignored, because it is useless information.

What is the error? Can you provide `package.json`/`package-json.lock`? Any additional information will help

**Comment by evilebottnawi** [comment]
> I thought it was only me but many of my team mates are experiencing the same thing.

They are use Windows?

**Comment by evilebottnawi** [comment]
+ You can put `console.log`/debugger here https://github.com/paulmillr/chokidar/blob/master/index.js#L541 and look on your event

**Comment by kane-mason** [comment]
> * You can put `console.log`/debugger here https://github.com/paulmillr/chokidar/blob/master/index.js#L541 and look on your event

nothing logs which i guess means it breaks before emitting anything

**Comment by kane-mason** [comment]
> > I thought it was only me but many of my team mates are experiencing the same thing.
> 
> They use Windows?
> 
> Do you use `npm`/`yarn`?

We use npm and the team is on a combination of mac and windows - the problem exists for both OS's

**Comment by kane-mason** [comment]
> No need 👍 please, it is not help, and will be ignored, because it is useless information.
> 
> What is the error or no errors? Can you provide `package.json`/`package-json.lock`? Any additional information will help
There is no error in the console, the process just exits. The `beforeExit` event is emitted from the `process` which according to the nodejs documentation means there is no more work scheduled and not from conditions causing explicit termination?
[package-lock.zip](https://github.com/webpack/watchpack/files/4854000/package-lock.zip)

**Comment by evilebottnawi** [comment]
@kane-mason nothing strange, can you try https://github.com/webpack/watchpack/issues/165#issuecomment-651814555? Without reproducible steps I can't help, sorry

**Comment by kane-mason** [comment]
> @kane-mason nothing strange, can you try [#165 (comment)](https://github.com/webpack/watchpack/issues/165#issuecomment-651814555)? Without reproducible steps I can't help, sorry

I did try see [comment](https://github.com/webpack/watchpack/issues/165#issuecomment-651990403)

Will try work more on a reproducible test case. Is there an easy way for us to opt out of using chokidar3?

**Comment by evilebottnawi** [comment]
We need to fix problem, `chokidar3` fixs more problems

**Comment by evilebottnawi** [comment]
Please provide full scripts how you run webpack, I think you have problems there

**Comment by kane-mason** [comment]
Ok i figured it out.

To cut a very long story short, we had an unmaintaned dependency `gulp-xml2json` which through a series of requires resulted in an old version of `graceful-fs` being required. This version `2.0.3` monkey patches the native `fs` module, and the 'new' `readdir` does not take into account the second `options` parameter which resulted in the callback being confused for this parameter, and never being called. Hence the watch actually never starts.

![image](https://user-images.githubusercontent.com/23186829/86351860-34191800-bc65-11ea-8d91-91bb29260e75.png)
![image](https://user-images.githubusercontent.com/23186829/86353302-5449d680-bc67-11ea-90de-70145fc3ca37.png)


So in summary, this project or more specifically `chokidar`, will run into silent annoying problems, if `graceful-fs` <= `2.0.3` is required **anywhere** else in your project.

**Comment by evilebottnawi** [comment]
@kane-mason It’s great that we caught this problem. I think we should leave it open to other developers. This is one of many reasons why we want to migrate from `chokidar` to native watch. 

Native performance and and compatibility issues must be addressed to `node` side. In long term developers should use native modules without any additional layers. Just my opinion.

**Comment by alexander-akait** [comment]
Closing due to inactivity. Thanks!
